modules/plugin/geopkg/src/main/java/org/geotools/geopkg/mosaic/GeoPackageReader.java

```diff
@@ -220,56 +244,132 @@ public GridCoverage2D read(String coverageName, GeneralParameterValue[] paramete
+                    if (newRes < resolution) {
+                        resolution = newRes;
+                        bestMatrix = matrix;
+                    }
+                }
+            }
+            
+            if (bestMatrix == null) {
+                // it means no level has tiles, return null, it's ok to do so
+                return null;
             }
 
             //take available tiles from database
-            leftTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), false, false);
-            rightTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), true, false);
-            topTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), false, true);
-            bottomTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), true, true);
-
-            double resX = (crs.getCoordinateSystem().getAxis(0).getMaximumValue() - crs.getCoordinateSystem().getAxis(0).getMinimumValue()) / bestMatrix.getMatrixWidth();
-            double resY = (crs.getCoordinateSystem().getAxis(1).getMaximumValue() - crs.getCoordinateSystem().getAxis(1).getMinimumValue()) / bestMatrix.getMatrixHeight();
-            double offsetX = crs.getCoordinateSystem().getAxis(0).getMinimumValue();
-            double offsetY = crs.getCoordinateSystem().getAxis(1).getMinimumValue();
-
-            if (requestedEnvelope != null) { //crop tiles to requested envelope                   
-                leftTile = Math.max(leftTile, (int) Math.round(Math.floor((requestedEnvelope.getMinimum(0) - offsetX) / resX )));
-                topTile = Math.max(topTile, (int) Math.round(Math.floor((requestedEnvelope.getMinimum(1) - offsetY) / resY )));
-                rightTile = Math.max(leftTile, (int) Math.min(rightTile, Math.round(Math.floor((requestedEnvelope.getMaximum(0) - offsetX) / resX ))));
-                bottomTile = Math.max(topTile, (int) Math.min(bottomTile, Math.round(Math.floor((requestedEnvelope.getMaximum(1) - offsetY) / resY ))));
-            } 
+            boundsLeftTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), false, false);
+            boundsRightTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), true, false);
+            boundsTopTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), false, true);
+            boundsBottomTile = file.getTileBound(entry, bestMatrix.getZoomLevel(), true, true);
+
+            Envelope entryBounds = entry.getTileMatrixSetBounds();
+            double resX = bestMatrix.getXPixelSize() * bestMatrix.getTileWidth();
+            double resY = bestMatrix.getYPixelSize() * bestMatrix.getTileHeight();
+            /*
+             * From the specification: "The tile coordinate (0,0) always refers to the tile in the upper left corner of the tile matrix at any zoom
+             * level, regardless of the actual availability of that tile."
+             * So remember the y axis goes from top to bottom, not the other way around
+             */
```

Comment:
thanks for the copious comments @aaime - given that we have had issues here before